### PR TITLE
Implement machine learning user model changes for Brave Ads

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/8739

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

Visit web sites and confirm if page classifications for Ads are as accurate as older versions of the app, or better

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 1]
```diff
@@ -511,6 +512,8 @@ void AdsImpl::OnTabClosed(
   destroy_info.tab_id = tab_id;
   const std::string report = reports.GenerateDestroyEventReport(destroy_info);
   ads_client_->EventLog(report);
+
+  // here
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: what's this placeholder for?

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 2]
```diff
@@ -788,26 +791,27 @@ std::string AdsImpl::ClassifyPage(
     const std::string& url,
     const std::string& content) {
   DCHECK(user_model_);
-  auto page_score = user_model_->ClassifyPage(content);
 
-  auto winning_category = GetWinningCategory(page_score);
+  const std::map<std::string, double> page_classification =
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: could we use `using CategoryScoreList = std::map<std::string, double>`. Or even more concrete `CategoryLikelihoodList` or `CategoryProbabilityList`?

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 2 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc]: Agreed to go with `CategoryProbabilityList`

**Comment by tmancey** [comment]
Browser tests failed for unrelated reasons to this PR

**Comment by tmancey** [comment]
iOS and macOS failed CI, restarting

**Comment by tmancey** [comment]
CI failed due to known rewards browser test issues

**Comment by tmancey** [comment]
CI failed on macOS, restarting macOS

**Comment by tmancey** [comment]
CI failed on Linux and macOS, restarting Linux and macoOS

**Comment by tmancey** [comment]
GitHub is showing CI failed, however checking CI, all passed

**Pull request merged**